[11081]: update workflow documentation (#33)
eclipse-velocitas/velocitas-docs: content/en/docs/about/deployment_model/vehicle_app_releases/_index.md

The Velocitas project provides a two-stage process for development, continuous integration, and release of a new version of a Vehicle App.

- **Stage 1 - Build & Test**

>   On every new push to the `main` branch or every update to Pull Request, a GitHub workflow is automatically executed to build your application as a container (optionally for different platforms), runs automated tests and code quality checks, and stores all results as GitHub artifacts for future reference with the default retention period of 90 days.

Reviewer: ```suggestion
  On every new push to the `main` branch or every update to a pull request, a GitHub workflow is automatically executed to build your application as a container (optionally for different platforms), runs automated tests and code quality checks, and stores all results as GitHub artifacts for future reference with a default retention period of 90 days.
```

Reviewer: What about the generated container: Is that stored as a build artefact or is it stored in ghcr?
In latter case: Is retention period the same?

Author: The release workflow pushes the images to the container registry and just use the created artifacts from the CI workflow.
Images on the container registry have no retention period. They have to be removed manually or even to define a cleaning up strategy on the organization level.

> 

Author: Done

  The workflow provides quick feedback during development and improves efficient collaboration.

- **Stage 2 - Release**
  Once the application is ready to be released in a new version, a dedicated release workflow is automatically executed as soon as you create a new release via GitHub.

>   The release workflow bundles all relevant artifacts into one tagged set of files and makes it possible to push this information to the preferred container registry of your choice (e.g. to be used for over-the-air update container registry) or use the information for quality assurance and documentation.

Reviewer: ```suggestion
  The release workflow bundles all relevant artifacts into one tagged set of files and makes it possible to push this information to the preferred container registry of your choice (e.g. to be used for an over-the-air update container registry) or use the information for quality assurance and documentation.
```

The drawing below illustrates the different workflows, actions and artifacts that are automatically created for you. Both workflows are intended as a sensible baseline and can be extended and adapted to your own project's needs.

![Project Overview](./publish_container.png)

## CI Workflow (ci.yml)

> The `Continuous Integration (CI) workflow` is triggered on every commit to the main branch or when creating/updating a Pull Request and contains a set of different actions that cover:

Reviewer: ```suggestion
The `Continuous Integration (CI) workflow` is triggered on every commit to the main branch or when creating/updating a pull request and contains a set of different actions that cover:
```

> The `Continuous Integration (CI) workflow` is triggered on every commit to the main branch or when creating/updating a pull request and contains a set of different actions that cover:

Reviewer: @emsameen: How do you think of this:
```suggestion
The `Continuous Integration (CI) workflow` is triggered on every commit to the main branch or when creating/updating a pull request and contains a set of actions to achieve the following objectives:
```

> - **Building a container for the app** - actions create a containerized version of the Vehicle App, the actions also support creating an image for multiple platforms and CPU architectures.

Reviewer: Is really multiple actions mentioned here or just a single action?
(Asking because of using the plural here)

Author: They are set of sequential actions to build, scan and publish the container images.

Reviewer: Hmh , I still find this a bit confusing because:
1. We are talking here about a GitHub workflows and the term "action(s)" may trigger thinking of a GitHub action, what this is not. It's jobs and steps. So, maybe another term would be less misleading, e.g. "activity", "task", or "procedure".
2. More important: In the sentence above this enumeration list, it's stated that the CI workflow "[...] contains a set of different **actions** that cover:" and then in the list itself, where I would expect that each single **action** is enumerated, we still speak of "actions" (in plural).

Author: I think the sentences was writing already before and I didn't change them here, I just updated some but I didn't invent that. 
From the other side, it is really multiple (actions/tasks/activities, whatever you call them) also called within the single stage. 
Please look to the [Build Image] (https://github.com/eclipse-velocitas/vehicle-app-python-template/blob/main/.github/workflows/build-docker-image.yml) sub-workflow which also internally has multiple activities. This workflow also called from the CI workflow.
I don't really see any objection with that but please feel free to have any suggestions and I will take them :)
- Just let me what do I need to change :) 
Thanks!

> - **Scanning for vulnerabilities** - actions scan your code and container for vulnerabilities and in case of findings the workflow will be marked as "failed".

Author: No, I think the scanning here is not the same as the code scanning in the CodeQL.
- Here the container image is scanned from vulnerabilities via some Github action after image creation. --> Again, I just updated the behaviour and not the context :)

- **Running unit tests & code coverage** - actions run unit tests and calculate code coverage for your application, in case of errors or unsatisfactory code coverage, the workflow will be marked as "failed".
- **Running integration tests** - actions provision a runtime instance and deploy all required services as containers together with your containerized application to allow for automatically executing integration test cases. In case the test cases fail, the workflow will be marked as "failed".

> - **Storing scan & test results as GitHub action artifacts** - actions store results from the previously mentioned actions for further reference or download as Github Action Artifacts.

Reviewer: Is using plural ("actions") correct?

Author: Yes, as above.

The Velocitas project provides a two-stage process for development, continuous integration, and release of a new version of a Vehicle App.

- **Stage 1 - Build & Test**
  On every new push to the `main` branch or every update to a pull request, a GitHub workflow is automatically executed to build your application as a container (optionally for different platforms), runs automated tests and code quality checks, and stores all results as GitHub artifacts for future reference with a [default retention period of 90 days](https://docs.github.com/en/organizations/managing-organization-settings/configuring-the-retention-period-for-github-actions-artifacts-and-logs-in-your-organization).

  The workflow provides quick feedback during development and improves efficient collaboration.

- **Stage 2 - Release**
  Once the application is ready to be released in a new version, a dedicated release workflow is automatically executed as soon as you create a new release via GitHub.

>   The release workflow bundles all relevant images and artifacts into one tagged set of files and makes it possible to be pushed to the GitHub Container Registry as well as for the information to be used for quality assurance and documentation accordingly.

Reviewer 2: ```suggestion
  The release workflow bundles all relevant images and artifacts into one tagged set of files and push it to the GitHub Container Registry. In addition, it makes the information ready to be used for quality assurance and documentation accordingly. The image in GitHub Container Registry can be used in your preferred  Over-The-Air (OTA) update system of your choice.
```

> The set of actions included in the `Release workflow` cover:

Reviewer: ```suggestion
The set of actions included in the `Release workflow` cover the objectives:
```

- **Generating and publishing QA information** - actions load the QA information from GitHub artifacts stored for the same commit reference and verify it. Additionally, release documentation is generated and added to the GitHub release. If there is no information available for the current commit, the release workflow will fail.

> - **Pull & label container image** - actions pull the Vehicle App container image based on the current commit hash from the GitHub Artifacts and label it with the specified tag version. If the image cannot be found, the workflow will fail.

Reviewer: ```suggestion
- **Pull & label container image** - actions pull the Vehicle App container image based on the current commit hash from the GitHub artifacts and label it with the specified tag version. If the image cannot be found, the workflow will fail.
```
